stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -17,51 +18,115 @@ namespace math {
+ * @tparam ContainerGradients tuple of any container gradients (Eigen types)
  */
-class precomputed_gradients_vari : public vari {
+template <typename ContainerOperands = std::tuple<>,
+          typename ContainerGradients = std::tuple<>>
+class precomputed_gradients_vari_template : public vari {
  protected:
   const size_t size_;
   vari** varis_;
   double* gradients_;
+  static_assert(std::tuple_size<ContainerOperands>::value
+                    == std::tuple_size<ContainerGradients>::value,
+                "precomputed_gradients_vari: ContainerOperands and "
+                "ContainerGradients should have same size!");
+  static constexpr size_t N_containers
+      = std::tuple_size<ContainerOperands>::value;
+  ContainerOperands container_operands_;
+  ContainerGradients container_gradients_;
+
+  /**
+   * Checks that sizes of containers in container operands and
+   * container_gradients match.
+   *
+   * @throws std::invalid_argument sizes do not match
+   */
+  template <size_t... Is>
+  void check_sizes(std::index_sequence<Is...>) {
+    static_cast<void>(std::initializer_list<int>{
+        (check_matching_dims("precomputed_gradients_vari", "operands",
+                             std::get<Is>(container_operands_), "gradients",
+                             std::get<Is>(container_gradients_)),
+         0)...});
+  }
 
  public:
   /**
    * Construct a precomputed vari with the specified value,
-   * operands, and gradients.
+   * operands, gradients and optionally container operands and containers of
+   * gradients.
    *
+   * @tparam ContainerOps tuple of any container operands (var_value
+   * containing Eigen types)
+   * @tparam ContainerGrads tupleof any container gradients (Eigen types)
    * @param[in] val The value of the variable.
    * @param[in] size Size of operands and gradients
    * @param[in] varis Operand implementations.
    * @param[in] gradients Gradients with respect to operands.
+   * @param container_operands any container operands
+   * @param container_gradients any container gradients
    */
-  precomputed_gradients_vari(double val, size_t size, vari** varis,
-                             double* gradients)
-      : vari(val), size_(size), varis_(varis), gradients_(gradients) {}
+  template <typename... ContainerOps, typename... ContainerGrads>
+  precomputed_gradients_vari_template(
+      double val, size_t size, vari** varis, double* gradients,
+      const std::tuple<ContainerOps...>& container_operands = std::tuple<>(),
+      const std::tuple<ContainerGrads...>& container_gradients = std::tuple<>())
+      : vari(val), size_(size), varis_(varis), gradients_(gradients) {
+    index_apply<N_containers>([&, this](auto... Is) {
+      this->container_operands_
+          = std::make_tuple(to_AD_stack(std::get<Is>(container_operands))...);
+      this->container_gradients_
+          = std::make_tuple(to_AD_stack(std::get<Is>(container_gradients))...);
+    });
```

Comment:
[optional] if we add `for_each` this can be

```suggestion
      : vari(val), size_(size), varis_(varis), gradients_(gradients), 
        container_operands_(for_each([](const auto& x) { return to_AD_stack(x);}, container_operands)),
        container_operands_(for_each([](const auto& x) { return to_AD_stack(x);}, container_gradients)),
```

Reply:
Not a major difference in my opinion. Fell free to change it if you decide to add it.